datafusion-examples/examples/advanced_udaf.rs

```diff
@@ -194,12 +210,196 @@ fn create_context() -> Result<SessionContext> {
     Ok(ctx)
 }
 
+struct GeometricMeanGroupsAccumulator<F>
+where
+    F: Fn(
+            <Float64Type as ArrowPrimitiveType>::Native,
+            u64,
+        ) -> Result<<Float64Type as ArrowPrimitiveType>::Native>
+        + Send,
+{
+    /// The type of the internal sum
+    prod_data_type: DataType,
+
+    /// The type of the returned sum
+    return_data_type: DataType,
+
+    /// Count per group (use u64 to make UInt64Array)
+    counts: Vec<u64>,
+
+    /// product per group, stored as the native type
+    prods: Vec<f64>,
+
+    /// Track nulls in the input / filters
+    null_state: NullState,
+
+    /// Function that computes the final geometric mean (value / count)
+    geo_mean_fn: F,
```

Comment:
I think the example would be simpler if you removed the generics and simply inlined the definition of `geo_mean_fn` into the callsite in `evaluate`.  The generics are needed for GroupsAccumulators that are specialized on type (e.g. a special one for Float32, Float64, etc).